File: app/src/main/java/de/dreier/mytargets/adapters/BackupAdapter.java
package de.dreier.mytargets.adapters;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;

import de.dreier.mytargets.R;
import de.dreier.mytargets.databinding.ItemImageDetailsSecondaryActionBinding;
import de.dreier.mytargets.utils.backup.BackupEntry;

public class BackupAdapter extends RecyclerView.Adapter<BackupAdapter.BackupViewHolder> {

    private final OnItemClickListener<BackupEntry> primaryActionListener;
    private final OnItemClickListener<BackupEntry> secondaryActionListener;
    private LayoutInflater inflater;
    private DateFormat formatDateTime;
    private List<BackupEntry> backupEntries = new ArrayList<>();

    public BackupAdapter(Context context, OnItemClickListener<BackupEntry> primaryActionListener, OnItemClickListener<BackupEntry> secondaryActionListener) {
        this.inflater = LayoutInflater.from(context);
        this.primaryActionListener = primaryActionListener;
        this.secondaryActionListener = secondaryActionListener;
        this.formatDateTime = SimpleDateFormat.getDateTimeInstance();
    }

    @Override
    public BackupViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = inflater.inflate(R.layout.item_image_details_secondary_action, parent, false);
        return new BackupViewHolder(v);
    }

    @Override
    public void onBindViewHolder(BackupViewHolder holder, int position) {
        BackupEntry p = backupEntries.get(position);
        final String modified = formatDateTime.format(p.getModifiedDate());
        holder.binding.name.setText(modified);
        holder.binding.details.setText(p.getHumanReadableSize());
        holder.binding.primaryAction
                .setOnClickListener(view -> primaryActionListener.onItemClicked(p));
        holder.binding.secondaryAction
                .setOnClickListener(view -> secondaryActionListener.onItemClicked(p));
    }

    @Override
    public int getItemCount() {
        return backupEntries.size();
    }

    public void setList(List<BackupEntry> list) {
        this.backupEntries = list;
        notifyDataSetChanged();
    }

    public void remove(BackupEntry backupEntry) {
        int index = backupEntries.indexOf(backupEntry);
        if (index > -1) {
            backupEntries.remove(index);
            notifyItemRemoved(index);
        }
    }

    public interface OnItemClickListener<T> {
        void onItemClicked(T item);
    }

    public static class BackupViewHolder extends RecyclerView.ViewHolder {

        public final ItemImageDetailsSecondaryActionBinding binding;

        public BackupViewHolder(View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
        }
    }
}

File: app/src/main/java/de/dreier/mytargets/features/settings/backup/BackupAdapter.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.features.settings.backup

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import de.dreier.mytargets.R
import de.dreier.mytargets.databinding.ItemDetailsSecondaryActionBinding
import org.threeten.bp.ZoneOffset
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class BackupAdapter(
    context: Context,
    private val primaryActionListener: OnItemClickListener<BackupEntry>,
    private val secondaryActionListener: OnItemClickListener<BackupEntry>
) : RecyclerView.Adapter<BackupAdapter.BackupViewHolder>() {
    private val inflater = LayoutInflater.from(context)
    private val formatDateTime = SimpleDateFormat.getDateInstance(DateFormat.LONG)
    private var backupEntries: MutableList<BackupEntry> = ArrayList()

    init {
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BackupViewHolder {
        val v = inflater.inflate(R.layout.item_details_secondary_action, parent, false)
        return BackupViewHolder(v)
    }

    override fun onBindViewHolder(holder: BackupViewHolder, position: Int) {
        val p = backupEntries[position]
        val modified = formatDateTime.format(p.lastModifiedAt)
        holder.binding.name.text = modified
        holder.binding.details.text = p.humanReadableSize
        holder.binding.primaryAction
            .setOnClickListener { primaryActionListener.onItemClicked(p) }
        holder.binding.secondaryAction
            .setOnClickListener { secondaryActionListener.onItemClicked(p) }
    }

    override fun getItemId(position: Int): Long {
        return backupEntries[position].lastModifiedAt
    }

    override fun getItemCount(): Int {
        return backupEntries.size
    }

    fun setList(list: MutableList<BackupEntry>) {
        this.backupEntries = list
        notifyDataSetChanged()
    }

    fun remove(backupEntry: BackupEntry) {
        val index = backupEntries.indexOf(backupEntry)
        if (index > -1) {
            backupEntries.removeAt(index)
            notifyItemRemoved(index)
        }
    }

    class BackupViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val binding = ItemDetailsSecondaryActionBinding.bind(itemView)
    }
}

interface OnItemClickListener<in T> {
    fun onItemClicked(item: T)
}
